src/r1cs.rs

```diff
@@ -347,14 +355,19 @@ impl<G: Group> R1CSShape<G> {
     // check if the number of variables are as expected, then
     // we simply set the number of constraints to the next power of two
     if self.num_vars == m {
-      return R1CSShape {
+      let mut r1cs_shape = R1CSShape {
         num_cons: m,
         num_vars: m,
         num_io: self.num_io,
         A: self.A.clone(),
         B: self.B.clone(),
         C: self.C.clone(),
+        digest: G::Scalar::ZERO,
       };
+
+      // it's a shame we have to recompute everything here; perhaps we can optimize this somehow
```

Comment:
We will indeed need that optimization, because the same bytes will be hashed again in https://github.com/lurk-lab/arecibo/blob/c3608280054fd183c65873b88251e2710d483c3f/src/lib.rs#L203-L204

Evidence that this is performance-sensitive exists in https://github.com/microsoft/Nova/pull/182/files, which introduced the benchmark and tests for the public parameter's digest.

Reply:
Can we compute it only if it's equal to `G1::Scalar::ZERO` (the default initial value)?

Reply:
Padding occurs as a one time cost at the end when producing the snark. Technically we currently don't even need to check any digest at this point, since it's assumed that the public params are already loaded and valid (though in the future it might be necessary to check again due to some new complexity). So my inclination to to delete this line and add a comment making it clear that this is "wrong," but we don't care.

Indeed, this is pretty compute heavy so we have to be careful.

Reply:
@arthurpaulino we could compute it for any default initial value (like `37`, or `None` with `Option<F>`). We just had to pick a default one to be able to serialize the structure (and make sure it's the same default each time).